contributors/design-proposals/storage/container-storage-interface-skip-attach.md

```diff
@@ -0,0 +1,81 @@
+
+Author: @jsafrane
+
+## Goal
+* Non-attachable CSI volumes should not require external attacher and `VolumeAttachment` instance creation. This will speed up pod startup.
+
+## Motivation
+Currently, CSI requires admin to start external CSI attacher for **all** CSI drivers, including those that don't implement attach/detach operation (such as NFS or all ephemeral Secrets-like volumes). Kubernetes Attach/Detach controller always creates `VolumeAttachment` objects for them and always waits until they're reported as "attached" by external CSI attacher.
+
+We want to skip creation of `VolumeAttachment` objects in A/D controller for CSI volumes that don't require 3rd party attach/detach.
+
+## Dependencies
+In order to skip both A/D controller attaching a volume and kubelet waiting for the attachment, both of them need to know if a particular CSI driver is attachable or not. In this document we expect that proposal #2514 is implemented and both A/D controller and kubelet has informer on `CSIDriver` so they can check if a volume is attachable easily.
+
+## Design
+### Volume plugins
+
+#### CSI volume plugin
+* Rework [`ProbeVolumePlugins`](https://github.com/kubernetes/kubernetes/blob/43f805b7bdda7a5b491d34611f85c249a63d7f97/pkg/volume/csi/csi_plugin.go#L58) to accept informer that watches CSIDriver. The plugin will store the informer for later.
+* Rework `Attach`, `Detach`, `VolumesAreAttached` and `WaitForAttach` to check for `CSIDriver` instance using the informer.
+	* If CSIDriver for the driver exists and it's attachable, perform usual logic.
+	* If CSIDriver for the driver exists and it's not attachable, return success immediately (basically NOOP).
+	* If CSIDriver for the driver does not exist, perform usual logic (i.e. treat the volume as attachable).
+	  * This keeps the behavior the same as in old Kubernetes version without CSIDriver object.
+	  * This also happens when CSIDriver informer has not been quick enough. It is suggested that CSIDriver instance is created **before** any pod that uses corresponding CSI driver can run.
+	    * In case that CSIDriver informer (or user) is too slow, CSI volume plugin `Attach()` will create `VolumeAttachment` instance and wait for (non-existing) external attacher to fulfill it. The CSI plugin shall recover when `CSIDriver` instance is created and skip attach. Any `VolumeAttachment` instance created here will be deleted on `Detach()`, see the next bullet.
+* In addition to the above, `Detach()` removes `VolumeAttachment` instance even if the volume is not attachable. This deletes `VolumeAttachment` instances created by old A/D controller or before `CSIDriver` instance was created.
+
+### A/D controller
+* A/D controller must pass a (shared) informer to watch CSIDriver and pass it to CSI volume plugin in [`ProbeVolumePlugins`](https://github.com/kubernetes/kubernetes/blob/8db5328c4c1f9467ab0d70ccb991a12d4675b6a7/cmd/kube-controller-manager/app/plugins.go#L82).
```

Comment:
I'll rephrase it, it does not need to be a/d controller, it can be kube-controller-manager itself. And I tried to add a parameter with the informer to `ProbeVolumePlugins()` and it's too early. kubelet does not have fully established API client at that time. So I'll either create a new non-shared informer in `Init()` of the plugin or add it to `VolumeHost`.